python-checks/src/main/java/org/sonar/python/checks/SklearnPipelineParameterAreCorrectCheck.java

```diff
@@ -0,0 +1,320 @@
+      .flatMap(expandRecursivePipelines());
+    a.forEach(
+      expressionAndPrefix -> getResult(expressionAndPrefix.tuple()).ifPresent(result1 -> out.put(expressionAndPrefix.prefix() + result1.stepName(), result1.classifierName())));
+    return out;
+  }
+
+  private static boolean isTwoElementTuple(List<Expression> elements) {
+    return elements.size() == 2;
+  }
+
+  private static ExpressionAndPrefix createEmptyExpressionAndPrefix(List<Expression> tuple) {
+    return new ExpressionAndPrefix(tuple, "");
+  }
+
+  private static Function<ExpressionAndPrefix, Stream<ExpressionAndPrefix>> expandRecursivePipelines() {
+    return expressionAndPrefix -> {
+      var tuple = expressionAndPrefix.tuple();
+      var step = tuple.get(0);
+      var classifier = tuple.get(1);
+
+      if (!step.is(Tree.Kind.STRING_LITERAL)) {
+        return Stream.of(expressionAndPrefix);
+      }
+
+      if (classifier.is(Tree.Kind.NAME)) {
+        var nestedPipelineCallExpressionOptional = Expressions.singleAssignedNonNameValue((Name) classifier)
+          .flatMap(TreeUtils.toOptionalInstanceOfMapper(CallExpression.class))
+          .filter(callExpression1 -> Optional.of(callExpression1)
+            .map(CallExpression::calleeSymbol).map(Symbol::fullyQualifiedName)
+            .filter("sklearn.pipeline.Pipeline"::equals)
+            .isPresent());
+
+        return nestedPipelineCallExpressionOptional.map(callExpression -> TreeUtils.nthArgumentOrKeyword(0, "steps", callExpression.arguments()))
+          .map(SklearnPipelineParameterAreCorrectCheck::getExpressionsFromArgument)
+          .orElse(Stream.empty())
+          .map(TreeUtils.toInstanceOfMapper(TupleImpl.class))
+          .filter(Objects::nonNull)
+          .map(TupleImpl::elements)
+          .filter(SklearnPipelineParameterAreCorrectCheck::isTwoElementTuple)
+          .map(elements -> new ExpressionAndPrefix(elements,
+            expressionAndPrefix.prefix() + ((StringLiteral) step).trimmedQuotesValue() + "__"))
+          .flatMap(expandRecursivePipelines());
+      }
+      return Stream.of(expressionAndPrefix);
+    };
+  }
+
+  private static Optional<Result> getResult(List<Expression> tuple) {
+    var step = tuple.get(0);
+    var classifier = tuple.get(1);
+    var stepName = Optional.ofNullable(step)
+      .flatMap(TreeUtils.toOptionalInstanceOfMapper(StringLiteral.class))
+      .map(StringLiteral::trimmedQuotesValue);
+    var classifierName = Optional.ofNullable(classifier)
+      .flatMap(TreeUtils.toOptionalInstanceOfMapper(CallExpression.class))
+      .map(CallExpression::calleeSymbol)
+      .filter(symbol -> symbol.is(Symbol.Kind.CLASS) && !Objects.equals(symbol.fullyQualifiedName(), "sklearn.pipeline.Pipeline"))
+      .map(ClassSymbol.class::cast);
+
+    return stepName.flatMap(stepName1 -> classifierName.map(classifierName1 -> new Result(stepName1, classifierName1)));
+  }
+
+  private record Result(String stepName, ClassSymbol classifierName) {
```

Comment:
Not convinced about the naming here. Maybe something more explicit on what should be expected in the tuple would be great. Maybe if you are facing issues with some naming, sometimes it could help namespacing them in classes that could hold some data and logic.